### DISCUSSION TITLE
Convert image quality to specified size？

### DISCUSSION THREAD

**Discussion by jishi92** [libvips/libvips#3258] [Q&A] [answered]
I want convert a jpeg image to specified size,such as 25KB.
Imagemagick supports this  feature.
Could libvips support it ?

**Comment by jcupitt** [comment]
Hi @jishi92,

You could implement this pretty easily in something like python. Just save to a string, then to a binary search while adjusting Q.

**Reply by jishi92** [reply]
Sorry,I'm not sure I understand what you mean.
I need traversal Q to adjust specified size?
Could you give me a code demo?

**Comment by jcupitt** [comment, marked as answer]
Just do a standard binary search. Something like:

```python
#!/usr/bin/python3

import sys
import pyvips

image = pyvips.Image.new_from_file(sys.argv[1])

lower = 0
upper = 100 
target = 100000
while upper - lower > 5:
    q = lower + (upper - lower) // 2
    data = image.write_to_buffer(".jpg", Q=q)
    print(f"q = {q}, size = {len(data)}")
    if len(data) > target:
        upper = q
    else:
        lower = q

with open(sys.argv[2], "wb") as f:
    f.write(data)

print(f"selected Q = {lower}, written to {sys.argv[2]}")
```

I see:

```
$ ./save-size.py ~/pics/shark.jpg x.jpg
q = 50, size = 45396
q = 75, size = 57106
q = 87, size = 76366
q = 93, size = 113089
q = 90, size = 96432
selected Q = 90, written to x.jpg
```

You might want to strip the JPG file as part of save.

**Reply by jishi92** [reply]
Thanks,I know.
So libvips adjust image size though adjust quality.
It should only get a close size, not an exact one.

**Reply by jishi92** [reply]
By the way, I think this binary search can Integrate into libvips and open a param to user,such as `sizeLimit`...

**Reply by jcupitt** [reply]
It's only a few lines of code and building it into libvips would not give a speedup, so I'm inclined to leave it out.
